File: YHDragContainer-OC/YHDragContainer/HorizontalDragViewController.m
//
//  HorizontalDragViewController.m
//  YHDragContainer
//
//  Created by apple on 2019/10/8.
//  Copyright © 2019 yinhe. All rights reserved.
//

#import "HorizontalDragViewController.h"
#import "YHDragCardContainer.h"

@interface HorizontalDragViewController () <YHDragCardDataSource, YHDragCardDelegate> {
    BOOL _disableDrag;
}
@property (nonatomic, strong) NSArray<NSString *> *models;
@property (nonatomic, strong) YHDragCardContainer *card;
@property (nonatomic, strong) UIBarButtonItem *reloadItem;
@property (nonatomic, strong) UIButton *revokeButton;
@property (nonatomic, strong) UIButton *nextButton;
@property (nonatomic, strong) UIButton *disableButton;
@property (nonatomic, strong) UIView *stateView;
@end

@implementation HorizontalDragViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    self.view.backgroundColor = [UIColor whiteColor];
    
    self.models = @[@"水星",
                    @"金星",
                    @"地球",
                    @"火星",
                    @"木星"];
    
    self.navigationItem.rightBarButtonItem = self.reloadItem;
    
    [self.view addSubview:self.card];
    [self.view addSubview:self.revokeButton];
    [self.view addSubview:self.nextButton];
    [self.view addSubview:self.disableButton];
    [self.view addSubview:self.stateView];
}

#pragma mark YHDragCardDataSource
- (int)numberOfCountInDragCard:(YHDragCardContainer *)dragCard{
    return (int)self.models.count;
}

- (UIView *)dragCard:(YHDragCardContainer *)dragCard indexOfCard:(int)index{
    UILabel *label = [[UILabel alloc] init];
    label.font = [UIFont boldSystemFontOfSize:50];
    label.textAlignment = NSTextAlignmentCenter;
    label.backgroundColor = [UIColor orangeColor];
    label.layer.cornerRadius = 5.0;
    label.layer.borderWidth = 1.0;
    label.layer.borderColor = [UIColor blackColor].CGColor;
    label.layer.masksToBounds = YES;
    label.text = [NSString stringWithFormat:@"%d -- %@", index, self.models[index]];
    return label;
}

#pragma mark YHDragCardDelegate
- (void)dragCard:(YHDragCardContainer *)dragCard didDisplayCard:(UIView *)card withIndex:(int)index{
    self.navigationItem.title = [NSString stringWithFormat:@"%d/%d", index + 1, (int)self.models.count];
}

- (void)dragCard:(YHDragCardContainer *)dragCard didSlectCard:(UIView *)card withIndex:(int)index{
    NSLog(@"点击卡片索引:%d", index);
}

- (void)dragCard:(YHDragCardContainer *)dragCard didRemoveCard:(UIView *)card withIndex:(int)index{
    NSLog(@"索引为%d的卡片滑出去了", index);
}

- (void)dragCard:(YHDragCardContainer *)dragCard didFinishRemoveLastCard:(UIView *)card{
    NSLog(@"最后一张卡片滑出去了");
    [self reload];
}

- (void)dragCard:(YHDragCardContainer *)dragCard currentCard:(UIView *)card withIndex:(int)index currentCardDirection:(YHDragCardDirection *)direction canRemove:(BOOL)canRemove{
    CGFloat ratio = ABS(direction.horizontalRatio) * 0.2 + 1.0;
    self.stateView.transform = CGAffineTransformMakeScale(ratio, ratio);
}

#pragma mark Action
// 刷新
- (void)reload{
    [self.card reloadData:YES];
}

// 撤销
- (void)revokeAction{
    [self.card revoke:YHDragCardDirectionTypeLeft]; // 从左边撤销
}

// 下一张
- (void)nextAction{
    [self.card nextCard:YHDragCardDirectionTypeRight]; // 从右边滑出去
}

// 禁用拖动&&启用拖动
- (void)disableDragAction{
    _disableDrag = !_disableDrag;
    [self setDisableDragButtonState];
    
    self.card.disableDrag = _disableDrag;
}

- (void)setDisableDragButtonState{
    NSString *title = @"";
    if (_disableDrag) {
        title = @"启用拖动手势";
    } else {
        title = @"禁用拖动手势";
    }
    [self.disableButton setTitle:title forState:UIControlStateNormal];
}
#pragma mark Getter
- (YHDragCardContainer *)card{
    if (!_card) {
        _card = [[YHDragCardContainer alloc] initWithFrame:CGRectMake(50, [UIApplication sharedApplication].statusBarFrame.size.height+44+40, self.view.frame.size.width - 100, 400)];
        _card.delegate = self;
        _card.dataSource = self;
        _card.minScale = 0.9;
        _card.removeDirection = YHDragCardRemoveDirectionHorizontal;
    }
    return _card;
}

- (UIBarButtonItem *)reloadItem{
    if (!_reloadItem) {
        _reloadItem = [[UIBarButtonItem alloc] initWithTitle:@"刷新" style:UIBarButtonItemStylePlain target:self action:@selector(reload)];
    }
    return _reloadItem;
}

- (UIButton *)revokeButton{
    if (!_revokeButton) {
        _revokeButton = [UIButton buttonWithType:UIButtonTypeSystem];
        [_revokeButton setTitle:@"撤销" forState:UIControlStateNormal];
        _revokeButton.backgroundColor = [UIColor grayColor];
        [_revokeButton setTitleColor:[UIColor whiteColor] forState:UIControlStateNormal];
        _revokeButton.frame = CGRectMake(30, self.card.frame.origin.y + self.card.frame.size.height + 40, 100, 40);
        [_revokeButton addTarget:self action:@selector(revokeAction) forControlEvents:UIControlEventTouchUpInside];
    }
    return _revokeButton;
}

- (UIButton *)nextButton{
    if (!_nextButton) {
        _nextButton = [UIButton buttonWithType:UIButtonTypeSystem];
        [_nextButton setTitle:@"下一张" forState:UIControlStateNormal];
        _nextButton.backgroundColor = [UIColor grayColor];
        [_nextButton setTitleColor:[UIColor whiteColor] forState:UIControlStateNormal];
        _nextButton.frame = CGRectMake([UIScreen mainScreen].bounds.size.width - 30.0 - 100.0, self.card.frame.origin.y + self.card.frame.size.height + 40, 100, 40);
        [_nextButton addTarget:self action:@selector(nextAction) forControlEvents:UIControlEventTouchUpInside];
    }
    return _nextButton;
}

- (UIButton *)disableButton{
    if (!_disableButton) {
        _disableButton = [UIButton buttonWithType:UIButtonTypeSystem];
        _disableButton.backgroundColor = [UIColor grayColor];
        [_disableButton setTitleColor:[UIColor whiteColor] forState:UIControlStateNormal];
        _disableButton.frame = CGRectMake(([UIScreen mainScreen].bounds.size.width - 120.0) / 2.0, self.card.frame.origin.y + self.card.frame.size.height + 40, 120, 40);
        [_disableButton addTarget:self action:@selector(disableDragAction) forControlEvents:UIControlEventTouchUpInside];
        [self setDisableDragButtonState];
    }
    return _disableButton;
}

- (UIView *)stateView{
    if (!_stateView) {
        _stateView = [[UIView alloc] initWithFrame:CGRectMake(([UIScreen mainScreen].bounds.size.width - 100.0) / 2.0, self.revokeButton.frame.origin.y+self.revokeButton.frame.size.height+40, 100, 100)];
        _stateView.backgroundColor = [UIColor purpleColor];
        _stateView.layer.cornerRadius = 100.0 / 2.0;
        _stateView.layer.masksToBounds = YES;
    }
    return _stateView;
}

@end

File: YHDragContainer/HorizontalDragViewController.swift
//
//  HorizontalDragViewController.swift
//  YHDragCard-Swift
//
//  Created by apple on 2019/9/30.
//  Copyright © 2019 yinhe. All rights reserved.
//

import UIKit
import YHDragCardSwift

class HorizontalDragViewController: UIViewController {
    
    private let models: [String] = ["水星",
                                    "金星",
                                    "地球",
                                    "火星",
                                    "木星",
                                    "土星",
                                    "天王星",
                                    "海王星",
                                    "木卫一",
                                    "土卫一"]
    
    private lazy var card: YHDragCard = {
        let card = YHDragCard(frame: CGRect(x: 50, y: UIApplication.shared.statusBarFrame.size.height + 44.0 + 40.0, width: self.view.frame.size.width - 100 , height: 400))
        card.dataSource = self
        card.delegate = self
        card.minScale = 0.9
        card.removeDirection = .horizontal
        return card
    }()
    
    lazy var reloadItem: UIBarButtonItem = {
        return UIBarButtonItem(title: "刷新", style: .plain, target: self, action: #selector(reload))
    }()
    
    lazy var revokeButton: UIButton = {
        let revokeButton = UIButton(type: .system)
        revokeButton.setTitle("撤销", for: .normal)
        revokeButton.backgroundColor = .gray
        revokeButton.setTitleColor(.white, for: .normal)
        revokeButton.frame = CGRect(x: 50, y: self.card.frame.origin.y + self.card.frame.size.height + 40, width: 100, height: 40)
        revokeButton.addTarget(self, action: #selector(revokeAction), for: .touchUpInside)
        return revokeButton
    }()
    
    lazy var nextButton: UIButton = {
        let nextButton = UIButton(type: .system)
        nextButton.setTitle("下一张", for: .normal)
        nextButton.backgroundColor = .gray
        nextButton.setTitleColor(.white, for: .normal)
        nextButton.frame = CGRect(x: UIScreen.main.bounds.size.width - 50.0 - 100.0, y: self.card.frame.origin.y + self.card.frame.size.height + 40, width: 100, height: 40)
        nextButton.addTarget(self, action: #selector(nextAction), for: .touchUpInside)
        return nextButton
    }()
    
    lazy var stateView: UIView = {
        let view = UIView()
        view.frame = CGRect(x: (UIScreen.main.bounds.size.width - 100.0)/2.0, y: self.revokeButton.frame.origin.y+self.revokeButton.frame.size.height+40, width: 100.0, height: 100.0)
        view.backgroundColor = UIColor.purple
        view.layer.cornerRadius = 100.0/2.0
        view.layer.masksToBounds = true
        return view
    }()
    
    override func viewDidLoad() {
        super.viewDidLoad()
        view.backgroundColor = .white
        navigationItem.rightBarButtonItem = reloadItem
        view.addSubview(self.card)
        view.addSubview(revokeButton)
        view.addSubview(nextButton)
        view.addSubview(stateView)
        
        // 请根据具体项目情况在合适的时机进行刷新
        self.card.reloadData(animation: false)
    }
}

extension HorizontalDragViewController{
    // 刷新
    @objc func reload() {
        self.card.reloadData(animation: true)
    }
    // 撤销
    @objc func revokeAction() {
        self.card.revoke(direction: .left)
    }
    // 下一张卡片
    @objc func nextAction() {
        self.card.nextCard(direction: .right)
    }
}



extension HorizontalDragViewController: YHDragCardDataSource {
    
    func numberOfCount(_ dragCard: YHDragCard) -> Int {
        return self.models.count
    }
    
    func dragCard(_ dragCard: YHDragCard, indexOfCell index: Int) -> YHDragCardCell {
        var cell = dragCard.dequeueReusableCell(withIdentifier: "ID") as? DemoCell
        if cell == nil {
            cell = DemoCell(reuseIdentifier: "ID")
        }
        let text = "索引：\(index)\n\n\(self.models[index])"
        cell?.set(title: text)
        return cell!
    }
}

extension HorizontalDragViewController: YHDragCardDelegate {
    func dragCard(_ dragCard: YHDragCard, didDisplayCell cell: YHDragCardCell, withIndexAt index: Int) {
        self.navigationItem.title = "当前卡片索引: \(index + 1)/\(self.models.count)"
    }
    
    func dragCard(_ dragCard: YHDragCard, didSelectIndexAt index: Int, with cell: YHDragCardCell) {
        let vc = NextViewController()
        self.navigationController?.pushViewController(vc, animated: true)
    }
    
    func dragCard(_ dragCard: YHDragCard, didRemoveCell cell: YHDragCardCell, withIndex index: Int, removeDirection: YHDragCardMoveDirection) {
        print("索引为\(index)的卡片滑出去了")
        switch removeDirection {
        case .up:
            print("向上移除")
        case .down:
            print("向下移除")
        case .left:
            print("向左移除")
        case .right:
            print("向右移除")
        case .none:
            print("default")
        }
    }
    
    func dragCard(_ dragCard: YHDragCard, didFinishRemoveLastCell cell: YHDragCardCell) {
        self.reload()
    }
    
    func dragCard(_ dragCard: YHDragCard, currentCell cell: YHDragCardCell, withIndex index: Int, currentCardDirection direction: YHDragCardDirection, canRemove: Bool) {
        let ratio = abs(direction.horizontalRatio) * 0.2
        self.stateView.transform = CGAffineTransform(scaleX: 1.0+ratio, y: 1.0+ratio)
    }
}

